### PR TITLE
Use refasm-only items for compilation

### PR DESCRIPTION
**Pull request by rainersigwald**
Use the new item provided by common targets to avoid recompiling a
project when a reference has changed only in implementation, not
interface.

Include a shim so that these targets can continue to work on older
MSBuild distributions. That keeps the NuGet-package-for-downlevel
scenario working, as well as making an easy transition period to the new
logic.

Requires Microsoft/msbuild#2039.

### CODE REVIEW

**Comment by rainersigwald** [comment]
@jcouv: implemented as discussed in https://github.com/Microsoft/msbuild/issues/1986#issuecomment-297547460.

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
+  </Target>
+
   <Target Name="CoreCompile"
           Inputs="$(MSBuildAllProjects);
                   @(Compile);
                   @(_CoreCompileResourceInputs);
                   $(ApplicationIcon);
                   $(AssemblyOriginatorKeyFile);
-                  @(ReferencePath);
+                  @(ReferencePathWithInterfaceOnlyAssemblies);
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets]: Consider "ReferencePathWithRefAssemblies".

File: src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets [hunk 2]
```diff
@@ -1,13 +1,25 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="ShimReferencePathsWhenCommonTargetsDoesNotUnderstandReferenceAssemblies"
+          BeforeTargets="CoreCompile"
+          Condition="'@(ReferencePathWithInterfaceOnlyAssemblies)' == ''">
+    <!-- Common targets should populate this item from 15.3, but this file
```
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/Microsoft.VisualBasic.Core.targets]: Nit: "dev15.3"

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets (hunk 1 above)
Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets]: Also, can you clarify the benefit of introducing a new name (why not keep "ReferencePath", but feed annotated content into it)?

Comment (jcouv) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets]: > Also, can you clarify the benefit of introducing a new name (why not keep "ReferencePath", but feed annotated content into it)?

Never mind, I see you already explained the options in https://github.com/Microsoft/msbuild/issues/1986#issuecomment-297547460 Thanks!

**Comment by jasonmalinowski** [comment]
I wasn't involved in the conversations of the ultimate decision to move to a new item type. @rainersigwald  are the motivations there documented and written up somewhere? The MSBuild item I see has some discussions but didn't quite give the backstory. I am very worried about project system and extensibility impact, and I want to make sure those were thought about.

**Comment by jcouv** [comment]
@jasonmalinowski I think https://github.com/Microsoft/msbuild/issues/1986#issuecomment-297547460 will answer your questions.

**Comment by rainersigwald** [comment]
That was my reasoning. It's entirely possible I've missed something important, though! What project-system/extensibility impact are you worried about?

**Comment by jasonmalinowski** [comment]
There are two impacts I'm concerned about:

1. Other targets that might exist in the world that might be generating ReferencePath items that are listed as "run before CoreCompile" -- the PR here has no protection against that. The 'shim' target might run _before_ such another target, and then those are broken. I'm not aware of stuff in the wild that does it, but I just wanted to raise it.
2. What places exist that consume ReferencePath? I know there's [this in Roslyn](https://github.com/dotnet/roslyn/blob/cebb8c48b934ad03ef24bc5bb3eca5605bd06cc8/src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs#L279-L282) and I see six references in the legacy project system. I don't know what exists in the new project system. Do we know if those need to be updated as well?

To be clear: I don't think this is the wrong approach...I just don't (personally) know if it's the right approach either. :smile:

File: build/scripts/build.ps1 [hunk 3]
```diff
@@ -25,7 +25,7 @@ function Print-Usage() {
 }
 
 function Run-Build() {
-    $buildArgs = "/v:m /m"
```
Comment (jcouv) [inline build/scripts/build.ps1]: Is this intended?

File: src/Compilers/Core/Portable/AdditionalTextFile.cs [hunk 4]
```diff
@@ -64,3 +64,5 @@ public override SourceText GetText(CancellationToken cancellationToken = default
         internal IList<DiagnosticInfo> Diagnostics => _diagnostics;
     }
 }
+
```
Comment (jcouv) [inline src/Compilers/Core/Portable/AdditionalTextFile.cs]: Please revert this file and the one below (empty changes). Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. I will try to produce a new signed build with this for end-to-end testing in VS tomorrow.

@jasonmalinowski Is there anyone else that we should pull in to review integration with project system?

File: build/scripts/build.ps1 [hunk 5]
```diff
@@ -25,7 +25,7 @@ function Print-Usage() {
...
+    $buildArgs = "/v:m /m /flp:v=diag"
```
Comment (jaredpar) [inline build/scripts/build.ps1]: This slows down our build considerably.  Why do we need to make this a part of the dev workflow?

**Comment by agocke** [comment]
I'm concerned about lack of tests here. How are we supposed to verify this change?

File: build/scripts/build.ps1 (hunk 5 above)
Comment (rainersigwald) [inline build/scripts/build.ps1]: You don't, local testing change escaped when I `git commit -am`-ed. Will revert.

**Comment by jasonmalinowski** [comment]
@dotnet/project-system can help review the project system impacts, as well as @dotnet/roslyn-ide who might have some knowledge of the legacy project system impacts.

**Comment by jcouv** [comment]
@agocke Yes, I ran into this problem in my own recent change for targets file :-( I did local verification then.
Maybe @rainersigwald would have some suggestions on validation strategy for targets files, since that seems a common msbuild problem.

That said, we're planning to produce a VS build with refout bits from msbuild and Roslyn, this week, so that we can do end-to-end validation. That will help cover.

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets [hunk 6]
```diff
@@ -1,13 +1,25 @@
...
 <!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
 <Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Target Name="ShimReferencePathsWhenCommonTargetsDoesNotUnderstandReferenceAssemblies"
+          BeforeTargets="CoreCompile"
+          Condition="'@(ReferencePathWithRefAssemblies)' == ''">
+    <!-- Common targets should populate this item from dev15.3, but this file
+         may be used (via NuGet package) on earlier MSBuilds. If the
+         adjusted-for-reference-assemblies item is not populated, just use
+         the older item's contents. -->
+    <ItemGroup>
```
Comment (tmeschter) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets]: If you're using a newer MSBuild is there any legitimate reason that `ReferencePathWithRefAssemblies` would be empty? If so are we breaking anything by forcing a value?

**Comment by rainersigwald** [comment]
@agocke Unfortunately, testing this kind of change in isolation is very difficult. MSBuild still runs a fairly large suite of (internal) tests that run builds of projects end-to-end and test for certain behaviors, and we should augment that test suite with some specific tests for ref-asm behavior. Does Roslyn have a similar suite, or are y'all dependent on others to catch build flow regressions? RPS scenarios also cover many build end-to-end configurations.

One major problem with the existing tests is that they work in VS-land, which requires having insertions of outside-the-VS-repo code. In order to run those tests, this change needs to be available in a Roslyn VSIX, so that it can be combined with an MSBuild VSIX built from the feature branch on our side and tested end-to-end. In MSBuild, we don't have a strict review policy on creating such a feature branch, getting a build from it, and testing it in VS--anyone on the core team can do that at their discretion. I don't know what the policies are in this repo.

This PR is what I'd like to see tested in that way--I would not be surprised if that testing revealed problems which will require followup, either here or in MSBuild. But I think the best way to gain confidence in this approach is to test it in VS.

File: src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets (hunk 6 above)
Comment (rainersigwald) [inline src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets]: The new common targets from Microsoft/msbuild#2039 unconditionally populate `@(ReferencePathWithRefAssemblies)` [here](https://github.com/Microsoft/msbuild/pull/2039/files#diff-02872751bc64e0ff785c9b6c1cd919b5R2033), so no--it shouldn't be empty unless `@(ReferencePath)` was too, in which case this should be harmless.

**Comment by jcouv** [comment]
Reviewers, any other feedback/questions/blockers?

**Comment by jasonmalinowski** [comment]
@jcouv Do we have the test plan written up somewhere?

**Comment by jcouv** [comment]
@jasonmalinowski Short answer is no, but I can work on that.
We have a placeholder test plan for compiler changes (https://github.com/dotnet/roslyn/issues/18612). Feel free to list suggestions there.
I have not thought systematically about end-to-end validation. We were just going to dogfood VS on Roslyn and MSBuild and also self-host (update our toolset).

Is there anything specific you'd like to see?

**Comment by jasonmalinowski** [comment]
@jcouv I'll add comments over there.

**Comment by agocke** [comment]
@rainersigwald @jcouv Is there any way we could get this into our bootstrap build before merging? Say, if we produced the MSBuild bits first and then updated Roslyn to consume them, then verified in the bootstrap build that we saw the expected behavior?

**Comment by agocke** [comment]
@jaredpar IIRC you added some validation to our bootstrap build to a build task before. Is that a viable option now?

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/18612

**Pull request merged**